umbraco/Umbraco-CMS
# Serialization and deserialziation with polymorphism

### Prerequisites


### Description
- Handling serialisation and deserialisation of interfaces when specifying `$type`
- Force Swagger to add `$type` when type implements interface

### Tests
- Try make endpoints that returns an interface or List of interfaces. Test that this now works
- Try make endpoints that receive an interface or List of interfaces. That that this now works
  -  Try posting types with `$type`
    -  Should deserialise correct  
  - Try posting types with unknown `$type`
    - Should be Bad Request 400 
  - Try posting without  `$type`
     - Should be Bad Request 400 
- Verify Swagger and the actual output is correct. 
  - `$type` should be required on types implementing an interface but not on all types.
### Appendix
This is a short test class I used.
```cs
using Microsoft.AspNetCore.Mvc;
using Umbraco.Cms.Api.Management.Controllers.Template;
using Umbraco.Cms.Core.Mapping;
using Umbraco.Cms.Core.Services;


namespace Umbraco.Cms.Api.Management.Controllers.Dictionary;

public class XXXController : TemplateControllerBase
{
    private readonly IDictionaryItemService _dictionaryItemService;
    private readonly IUmbracoMapper _umbracoMapper;

    public XXXController(IDictionaryItemService dictionaryItemService, IUmbracoMapper umbracoMapper)
    {
        _dictionaryItemService = dictionaryItemService;
        _umbracoMapper = umbracoMapper;
    }

    [HttpGet("polymorphism")]
    [MapToApiVersion("1.0")]
    public async Task<ActionResult<List<IItem>>> Polymorphism()
    {
        var result = new List<IItem>()
        {

            new Item1()
            {
                Name = "name 1",
                Title = "title"
            },
            new Item2()
            {
                Name = "name 2",
                Age = 42
            }
        };

        return await Task.FromResult(result);
    }

    [HttpPost("polymorphism")]
    [MapToApiVersion("1.0")]
    public async Task<ActionResult<List<IItem>>> PostPolymorphism(List<IItem> items)
    {
        if (!ModelState.IsValid)
        {
            return BadRequest();
        }

        return await Task.FromResult(items);
    }
}

public interface IItem
{
    public string Name { get; }
}


public class Item1 : IItem
{
    public required string Name { get; set; }
    public required string Title { get; set; }
}

public class Item2 : IItem
{
    public required string Name { get; set; }
    public required int Age { get; set; }
}


```
